Spigot-Server-Patches/0419-Ticket-level-per-player-view-distance-API.patch

```diff
@@ -0,0 +1,1150 @@
+-                PlayerChunk playerchunk = (PlayerChunk) objectiterator.next();
+-                ChunkCoordIntPair chunkcoordintpair = playerchunk.i();
++    //Paper start ticket level pp-view-dist
++    public void updateTrackViewDistance(EntityPlayer player, int i) {
++        for (int x = player.chunkX - i; x <= player.chunkX + i; ++x) {
++            for (int z = player.chunkZ - i; z <= player.chunkZ + i; ++z) {
++                ChunkCoordIntPair chunkcoordintpair = new ChunkCoordIntPair(x, z);
+                 Packet<?>[] apacket = new Packet[2];
++                int i1 = b(chunkcoordintpair, player, true);
++                boolean flag = i1 <= player.lastTrackViewDistance;
++                boolean flag1 = i1 <= i;
+ 
+-                this.a(chunkcoordintpair, false).forEach((entityplayer) -> {
+-                    int l = b(chunkcoordintpair, entityplayer, true);
+-                    boolean flag = l <= k;
+-                    boolean flag1 = l <= this.viewDistance;
+-
+-                    this.sendChunk(entityplayer, chunkcoordintpair, apacket, flag, flag1);
+-                });
++                this.sendChunk(player, chunkcoordintpair, apacket, flag, flag1);
+             }
+         }
+-
+     }
++    //Paper end
+ 
+     protected void sendChunk(EntityPlayer entityplayer, ChunkCoordIntPair chunkcoordintpair, Packet<?>[] apacket, boolean flag, boolean flag1) {
+         if (entityplayer.world == this.world) {
+@@ -966,10 +972,13 @@ public class PlayerChunkMap extends IChunkLoader implements PlayerChunk.d {
+                 PlayerChunk playerchunk = this.getVisibleChunk(chunkcoordintpair.pair());
+ 
+                 if (playerchunk != null) {
+-                    Chunk chunk = playerchunk.getChunk();
++                    Chunk chunk = playerchunk.getTickingChunk(); //for clarification
+ 
+                     if (chunk != null) {
+                         this.a(entityplayer, apacket, chunk);
++                    } else {
++                        //send unload for non ticking chunks (> 32)
++                        entityplayer.a(chunkcoordintpair); //Paper ticket level pp-view-dist
+                     }
+ 
+                     PacketDebug.a(this.world, chunkcoordintpair);
+@@ -1201,29 +1210,38 @@ public class PlayerChunkMap extends IChunkLoader implements PlayerChunk.d {
+     }
+ 
+     boolean isOutsideOfRange(ChunkCoordIntPair chunkcoordintpair, boolean reducedRange) {
+-        int chunkRange = world.spigotConfig.mobSpawnRange;
+-        chunkRange = (chunkRange > world.spigotConfig.viewDistance) ? (byte) world.spigotConfig.viewDistance : chunkRange;
+-        chunkRange = (chunkRange > 8) ? 8 : chunkRange;
+-
+-        final int finalChunkRange = chunkRange; // Paper for lambda below
+         //double blockRange = (reducedRange) ? Math.pow(chunkRange << 4, 2) : 16384.0D; // Paper - use from event
+         // Spigot end
+         long i = chunkcoordintpair.pair();
+-
```

Comment:
this cannot be in the view distance patch, move it to another patch

Reply:
already moved in rewrite